### PR TITLE
Fix Caffe2 ==> ONNX converter to handle three models

### PR DESCRIPTION
**Pull request by houseroad**
1) Handle legacy_pad
2) Remove unnecessary initializers on real input.

### CODE REVIEW

File: caffe2/python/onnx/frontend.py [hunk 1]
```diff
@@ -510,10 +544,18 @@ def ssa_name(name, version):
...
         if init_net:
-            for op in init_net.op:
-                assert re.match('GivenTensor.*Fill', op.type)
+            index = 0
+            while index < len(init_net.op):
+                op = init_net.op[index]
                 assert len(op.output) == 1
+                if op.output[0] in value_info:
+                    # Delete fake initializer on the real input
+                    fake_init = True
```
Comment (bddppq) [inline caffe2/python/onnx/frontend.py]: not used

File: caffe2/python/onnx/frontend.py [hunk 2]
```diff
@@ -510,10 +544,18 @@ def ssa_name(name, version):
...
+                    del init_net.op[index]
```
Comment (bddppq) [inline caffe2/python/onnx/frontend.py]: ssa step is not supposed to do this :-)

File: caffe2/python/onnx/frontend.py [hunk 3]
```diff
@@ -510,10 +544,18 @@ def ssa_name(name, version):
             return '{}_{}'.format(name, version)
 
         if init_net:
-            for op in init_net.op:
-                assert re.match('GivenTensor.*Fill', op.type)
+            index = 0
+            while index < len(init_net.op):
+                op = init_net.op[index]
                 assert len(op.output) == 1
+                if op.output[0] in value_info:
```
Comment (bddppq) [inline caffe2/python/onnx/frontend.py]: It's safer to whitelist couple operators types that can be deleted.

File: caffe2/python/onnx/frontend.py [hunk 4]
```diff
@@ -211,7 +212,37 @@ def apply_trans(k, dim=2, ks=None):
...
+            # So we use the explicit padding to replace legacy_pad.
+            # pad[end] = output_size[start + 2] * stride[start] - pad[start] - 1 + kernel[start] - input[start + 2]
+            # end = start + len(pad) / 2
+            assert node.op_type.endswith("Pool")
+            assert not node.op_type.startswith("Global")
+            input_size = shapes[node.input[0]]
+            output_size = shapes[node.output[0]]
+            assert len(output_size) == 4
+            if legacy_pad_attr.i == 1:
+                attrs.pop('pads')
```
Comment (bddppq) [inline caffe2/python/onnx/frontend.py]: `pads` and `legacy_pad` should not be set simultaneously:
https://github.com/caffe2/caffe2/blob/02ac8365935bcaf5f6552ba6d87f1285631a219b/caffe2/operators/conv_pool_op_base.h#L105

File: caffe2/python/onnx/frontend.py [hunk 5]
```diff
@@ -211,7 +212,37 @@ def apply_trans(k, dim=2, ks=None):
...
+            # The problem here is that, Pool op in Caffe may add an additional pixel, if the last part is smaller than stride.
+            # So we use the explicit padding to replace legacy_pad.
+            # pad[end] = output_size[start + 2] * stride[start] - pad[start] - 1 + kernel[start] - input[start + 2]
+            # end = start + len(pad) / 2
+            assert node.op_type.endswith("Pool")
+            assert not node.op_type.startswith("Global")
+            input_size = shapes[node.input[0]]
+            output_size = shapes[node.output[0]]
+            assert len(output_size) == 4
+            if legacy_pad_attr.i == 1:
```
Comment (bddppq) [inline caffe2/python/onnx/frontend.py]: Use the LegacyPadding.VALID instead

File: caffe2/python/onnx/frontend.py [hunk 6]
```diff
@@ -211,7 +212,37 @@ def apply_trans(k, dim=2, ks=None):
...
+                new_attr = make_attribute('auto_pad', 'VALID')
+                attrs[new_attr.name] = new_attr
+            elif legacy_pad_attr.i == 2:
+                # default behavior in Caffe2 is SAME_UPPER
+                # https://github.com/caffe2/caffe2/blob/master/caffe2/operators/conv_pool_op_base.h#L39
+                attrs.pop('pads')
+                new_attr = make_attribute('auto_pad', 'SAME_UPPER')
+                attrs[new_attr.name] = new_attr
+            elif legacy_pad_attr.i == 3:
+                print('Converting legacy padding to explict padding.')
```
Comment (bddppq) [inline caffe2/python/onnx/frontend.py]: nit: use logging

File: caffe2/python/onnx/frontend.py [hunk 7]
```diff
@@ -211,7 +212,37 @@ def apply_trans(k, dim=2, ks=None):
...
+                for i in range(2):
+                    attrs['pads'].ints[i + 2] = (output_size[i + 2] * attrs['strides'].ints[i] - attrs['pads'].ints[i]
```
Comment (bddppq) [inline caffe2/python/onnx/frontend.py]: Could you put the math equation as comment here?

File: caffe2/python/onnx/frontend.py [hunk 8]
```diff
@@ -211,7 +212,37 @@ def apply_trans(k, dim=2, ks=None):
         apply_trans('stride')
         apply_trans('dilation')
         apply_trans('adj')
-        apply_trans('pad', 4)
+        apply_trans('pad', dim=4)
+
+        legacy_pad_attr = attrs.pop('legacy_pad', None)
```
Comment (yinghai) [inline caffe2/python/onnx/frontend.py:218]: nit: make this a function?

File: caffe2/python/onnx/frontend.py (hunk 3 above)
Comment (houseroad) [reply caffe2/python/onnx/frontend.py]: ConstantFill and Given*TensorFill

**Review by houseroad — COMMENTED** [review]
Have addressed the comments

**Comment by houseroad** [comment]
@caffe2bot retest this please

File: caffe2/python/onnx/frontend.py (hunk 8 above)
Comment (yinghai) [reply caffe2/python/onnx/frontend.py:218]: Yes, it looks long enough... But it's just nit.

File: caffe2/python/onnx/frontend.py [hunk 9]
```diff
@@ -504,14 +539,28 @@ def _annotate_consumed(cls, graph_def):
                     consumes,
                 )])
 
+    @classmethod
+    def _filter_fake_init(cls, init_net, value_info):
+        if init_net:
+            index = 0
+            while index < len(init_net.op):
```
Comment (yinghai) [inline caffe2/python/onnx/frontend.py]: Python itself actually has a `filter` function. You can try using that to save some lines.

File: caffe2/python/onnx/tests/c2_ref_test.py [hunk 10]
```diff
@@ -352,6 +352,15 @@ def test_shufflenet(self):
     def test_densenet121(self):
         self._test_net('densenet121')
 
+    def test_bvlc_googlenet(self):
```
Comment (yinghai) [inline caffe2/python/onnx/tests/c2_ref_test.py:355]: Awesome!

File: caffe2/python/onnx/frontend.py (hunk 9 above)
Comment (houseroad) [reply caffe2/python/onnx/frontend.py]: Here, we cannot have assignment like `init_net.op[i] = new_op`. But applied something similar to make the code look nicer.

**Comment by yinghai** [comment]
@caffe2bot retest this please

File: caffe2/python/onnx/frontend.py [hunk 11]
```diff
@@ -504,14 +539,25 @@ def _annotate_consumed(cls, graph_def):
                     consumes,
                 )])
 
+    @classmethod
+    def _filter_fake_init(cls, init_net, value_info):
+        if init_net:
+            fake_inits = [op for op in init_net.op
+                          if len(op.output) == 1 and op.output[0] in value_info and
+                          re.match('GivenTensor.*Fill|ConstantFill', op.type)]
+            for fake_init in fake_inits:
```
Comment (yinghai) [inline caffe2/python/onnx/frontend.py:548]: You can do 
```
init_net.op = filter(lambda x: x not in fake_inits, init_net.op)
````
to replace everything from here.

**Review by yinghai — APPROVED** [review]
LGTM. Thanks for adding the tests!

File: caffe2/python/onnx/frontend.py (hunk 11 above)
Comment (houseroad) [reply caffe2/python/onnx/frontend.py:548]: The problem here is... you cannot assign anything to init_net.op. The only way I have figured out to change op, is the remove and del.

**Pull request merged**